README.md
<div align="center">
  <br>
  <img alt="Open Sauced" src="https://i.ibb.co/7jPXt0Z/logo1-92f1a87f.png" width="300px">
  <h1>🍕 Open Sauced Populate Supabase Test🍕</h1>
  <strong>The path to your next Open Source contribution</strong>
</div>
<br>
<p align="center">
  <img src="https://img.shields.io/github/languages/code-size/0-vortex/open-sauced-supabase-cron-test" alt="GitHub code size in bytes">
  <img src="https://img.shields.io/github/commit-activity/w/0-vortex/open-sauced-supabase-cron-test" alt="GitHub commit activity">
  <a href="https://github.com/0-vortex/open-sauced-supabase-cron-test/issues">
    <img src="https://img.shields.io/github/issues/0-vortex/open-sauced-supabase-cron-test" alt="GitHub issues">
  </a>
  <a href="https://github.com/0-vortex/open-sauced-supabase-cron-test/releases">
    <img src="https://img.shields.io/github/v/release/0-vortex/open-sauced-supabase-cron-test.svg?style=flat" alt="GitHub Release">
  </a>
  <a href="https://discord.gg/U2peSNf23P">
    <img src="https://img.shields.io/discord/714698561081704529.svg?label=&logo=discord&logoColor=ffffff&color=7389D8&labelColor=6A7EC2" alt="Discord">
  </a>
  <a href="https://twitter.com/saucedopen">
    <img src="https://img.shields.io/twitter/follow/saucedopen?label=Follow&style=social" alt="Twitter">
  </a>
</p>


## 📖 Prerequisites

In order to run the project we need the following software binaries installed on our development machines: 
- `node>=16.7.0`
- `npm>=8.0.0`
- `docker>=20.10.12`
- `supabase>=0.18.0`

# 🔑 Database commands

### Starting the Supabase Studio

First thing we have to do for [local development](https://supabase.com/docs/guides/local-development) is start the studio locally at [localhost:54321](http://localhost:54321):

```shell
npm run db:start
```

### Make changes

If we are adding a new table structure, first do it visually in the Supabase Studio and test locally.

Check the migration difference with the following command:

```shell
npm run db:changes
```

If everything is fine we can run the following command to apply the changes to the database:

```shell
npm run db:commit add_table_name
```

### Test changes

Simplest way to test the migrations are working is to reset the local database:

```shell
npm run db:reset
```

### Push changes

If everything is fine we can push the changes to the remote database:

```shell
npm run db:push
```

# 🔑 Supabase structure

Click the image to see the schema diagram full documentation.

[![er](./supabase/diagrams/schema.svg)](./supabase/diagrams/README.md)

# 🤝 Contributing

We encourage you to contribute to Open Sauced! Please check out the [Contributing guide](https://docs.opensauced.pizza/contributing/introduction-to-contributing/) for guidelines about how to proceed.

<img align="right" src="https://i.ibb.co/CJfW18H/ship.gif" width="200"/>

## 🍕 Community

Got Questions? Join the conversation in our [Discord](https://discord.gg/U2peSNf23P).  
Find Open Sauced videos and release overviews on our [YouTube Channel](https://www.youtube.com/channel/UCklWxKrTti61ZCROE1e5-MQ).

## 🎦 Repository Visualization

[![Visualization of this repository](./public/diagram.svg)
](./src)

## ⚖️ LICENSE

MIT © [Open Sauced](LICENSE)


CHANGELOG.md
# 📦 0-vortex/open-sauced-supabase-cron-test changelog

[![conventional commits](https://img.shields.io/badge/conventional%20commits-1.0.0-yellow.svg)](https://conventionalcommits.org)
[![semantic versioning](https://img.shields.io/badge/semantic%20versioning-2.0.0-green.svg)](https://semver.org)

> All notable changes to this project will be documented in this file

## [1.15.0](https://github.com/0-vortex/open-sauced-supabase-cron-test/compare/v1.14.1...v1.15.0) (2022-09-01)


### 🍕 Features

* add logged in supabase to github users backfill ([574d4bd](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/574d4bd12e8c4ab5c5c51748b8970e1c41b7a19e))


### 🐛 Bug Fixes

* remove concurrency from sql crawlers ([2cf1e0b](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/2cf1e0b1b313503de888ba489764de5f1f03f465))

### [1.14.1](https://github.com/0-vortex/open-sauced-supabase-cron-test/compare/v1.14.0...v1.14.1) (2022-08-20)


### 🐛 Bug Fixes

* correct new soft delete enabled stars tables ([d6eb82b](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/d6eb82b38c1273a8a97ab2ffc4d3623495ddbfcb))

## [1.14.0](https://github.com/0-vortex/open-sauced-supabase-cron-test/compare/v1.13.2...v1.14.0) (2022-08-05)


### 🍕 Features

* add stargazers and submissions to cron backups ([43c72a0](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/43c72a0521499cf4e34106d8c31a4398afd62612))

### [1.13.2](https://github.com/0-vortex/open-sauced-supabase-cron-test/compare/v1.13.1...v1.13.2) (2022-07-28)


### 🐛 Bug Fixes

* correct user stars breaking on suspended installations ([bf07286](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/bf07286ad5560bbe21b0d9581abfaf72b4dfbf47))

### [1.13.1](https://github.com/0-vortex/open-sauced-supabase-cron-test/compare/v1.13.0...v1.13.1) (2022-05-29)


### 🧑‍💻 Code Refactoring

* correct workflow step names ([781db39](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/781db394fd60e35622c17801cb54f0502685277a))

## [1.13.0](https://github.com/0-vortex/open-sauced-supabase-cron-test/compare/v1.12.2...v1.13.0) (2022-05-17)


### 🍕 Features

* add all contributors cron job ([b0bb698](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/b0bb69862f7afa5ae8c425b1afeaac42a7cb9476))

### [1.12.2](https://github.com/0-vortex/open-sauced-supabase-cron-test/compare/v1.12.1...v1.12.2) (2022-05-16)


### 🤖 Build System

* add current seeds to fix mkdir fallacy ([71fdeaf](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/71fdeaf4d972e08102dd836d167ca4c1553bcd77))

### [1.12.1](https://github.com/0-vortex/open-sauced-supabase-cron-test/compare/v1.12.0...v1.12.1) (2022-05-16)


### 🐛 Bug Fixes

* correct seed tables ([a304120](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/a3041201a30a529a5152866501b4e3bb692dbecb))

## [1.12.0](https://github.com/0-vortex/open-sauced-supabase-cron-test/compare/v1.11.1...v1.12.0) (2022-05-16)


### 🍕 Features

* make background contributor fetch cron work ([f2133e0](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/f2133e0b715a5240a8369a0a8dfe8fd47464a0da))

### [1.11.1](https://github.com/0-vortex/open-sauced-supabase-cron-test/compare/v1.11.0...v1.11.1) (2022-05-16)


### 🐛 Bug Fixes

* refresh cron job cacbe ([205ef16](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/205ef1651272903f58da9b3683515603eec0ffaa))

## [1.11.0](https://github.com/0-vortex/open-sauced-supabase-cron-test/compare/v1.10.1...v1.11.0) (2022-05-16)


### 🍕 Features

* update contributor minimal table structure ([46c7cc3](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/46c7cc31b3beeb19b9f7376be4eca50f1cfcecde))

### [1.10.1](https://github.com/0-vortex/open-sauced-supabase-cron-test/compare/v1.10.0...v1.10.1) (2022-05-16)


### 🐛 Bug Fixes

* correct some minor inconsistencies with new tables ([b6ce092](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/b6ce0924408773708653ac0f85e0f2c2b7c8a96b))

## [1.10.0](https://github.com/0-vortex/open-sauced-supabase-cron-test/compare/v1.9.8...v1.10.0) (2022-05-15)


### 🐛 Bug Fixes

* correct merge commits ([73ab78d](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/73ab78d2f80a51522adb1f931b3d898020510918))


### 🤖 Build System

* update deps to latest versions ([8cb8b53](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/8cb8b53b19d7124d2abf245e08208506c3035d75))
* update package manifest ([5a824a0](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/5a824a0c32e0e3b61e9516ecc9da70504e53b410))


### 🧑‍💻 Code Refactoring

* migrate scripts over to the new structure ([3aae0e1](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/3aae0e1b9ad73a93f8713c0d9add34a12cf6848a))


### 🍕 Features

* adjust tables to attach delayed contributors updates ([a222377](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/a222377825d5af0872dd3b5f68623df85afa0fac))
* replace populate script with new relationship db ([4950fba](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/4950fba7ab3068afe0c2ce8b629f5f6e6044238f))
* update database configuration ([c84c5a2](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/c84c5a2698869307eb572e1baaf0d9ad4aa8b29c))
* update seed scripts to new structure ([79fd3ef](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/79fd3ef78ddb7ce9443bb3110022ae00300583ef))
* wip contributors script and 100 users backfill ([07b81c6](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/07b81c68bd71cf28186456964cff2784a3ccf9ce))
* wip improve user stars into standalone repos ([17ba4ef](https://github.com/0-vortex/open-sauced-supabase-cron-test/commit/17ba4efbed38d9b19650bd7955466a2c6039e1ed))

### [1.9.8](https://github.com/0-vortex/open-sauced-supabase-cron-test/compare/v1.9.7...v1.9.8) (2022-03-25)


### 🐛 Bug Fixes


src/seed.js
import { p } from '@antfu/utils'

import { supabase, supaCount, supaDump, supaSeed } from './lib/supabase.js';
import consoleHeader from "./lib/consoleHeader.js";

const lastExecuted = new Date()
const tables = [
  'users',
  'repos',
  'contributions',
  'users_to_repos_stars',
  'users_to_repos_submissions',
  'users_to_repos_votes',
  'users_to_repos_stargazers',
]
const queue = []

consoleHeader(' OPEN |SAUCED', {
  font: 'block',
})
console.log(`Started execution at ${lastExecuted}`)

async function run() {
  consoleHeader('Parsing tables');
  await p(tables)
    .map(async (table) => {
      console.log(`Fetching counts for table ${table}`)
      const { count, error } = await supaCount(table)

      if (error) {
        console.log(`Error counting table ${table}`, error)
        return
      }

      console.log(`Table "${table}" count: ${count}`)

      if (count > 0) {
        queue.push({
          table,
          count,
        })
      }
    })

  consoleHeader('Parsing data')
  const seeds = await p(queue)
    .map(async ({ table, count }) => {
      console.log(`Parsing rows for table ${table}`)
      const tableQueue = p()
      const rows = []

      for (let i = 0; i < Math.ceil(count / 1000) * 1000; i += 1000) {
        tableQueue.add(
          supabase
            .from(table)
            .select('*')
            .range(i, i + 1000)
        )
      }

      await tableQueue
        .then((pages) => pages.forEach(({data, error}) => {
          !error && rows.push(...data)
        }))

      return {
        table,
        rows,
      }
    })
    .map(async ({ table, rows }) => {
      console.log(`Seeding ${table}.sql`)
      return supaDump('./seed/', table, Object.keys(rows[0]), rows)
    })

  consoleHeader('Versioning changes')
  await supaSeed(seeds)
  console.log('Wrote changes to supabase/seed.sql, make sure to commit this file')

  consoleHeader('Finished')
}

await run()


src/index.js
import { App } from 'octokit'
import { createClient } from '@supabase/supabase-js'
import { writeFile } from 'node:fs/promises'

import api from './lib/persistedGraphQL.js'
import fetchContributorNames from './lib/contributorNameHelper.js'
import consoleHeader from './lib/consoleHeader.js'
import cron from './cron.json'

const anon_key = process.env.SUPABASE_ANON_KEY
const supabaseUrl = process.env.SUPABASE_URL
const limitDays = parseInt(process.env.LIMIT_DAYS) || 2
let limitUsers = parseInt(process.env.LIMIT_USERS) || 5
const checked = {...cron.checked}
const lastExecuted = new Date()
const parsedCache = {}
const parseInstallations = []
const parseData = []

consoleHeader('OPEN SAUCED', {
  font: 'block',
});
console.log(`Started execution at ${lastExecuted}`)

for (const item in checked) {
  const date = new Date(checked[item].lastExecuted)
  const diff = lastExecuted - date

  // generate a cache of items with offset dates
  parsedCache[item] = {
    ...checked[item],
    offsetDays: Number(diff / 1000 / 60 / 60 / 24).toFixed(0)
  }
}

// Create a single supabase client for interacting with your database
const supabase = createClient(supabaseUrl, anon_key)

async function run() {
  const app = new App({
    appId: +process.env.OPEN_SAUCED_APP_ID,
    privateKey: process.env.OPEN_SAUCED_PRIVATE_KEY,
  })

  consoleHeader('Parsing installations');
  for await (const {installation} of app.eachInstallation.iterator()) {
    if (installation.account.login === 'open-sauced') {
      continue
    }

    console.log(`Fetching data for installation ${installation.account.login}`)
    // we don't have a local commit log of checking these installations
    if (typeof parsedCache[installation.account.login] === 'undefined') {
      console.log(`${installation.account.login} is not in our cache, adding to the queue`)
      parseInstallations.push(installation);
    } else {
      // we have a local cache of this installation
      let parsed = false;

      if (
        parsedCache[installation.account.login].offsetDays >= limitDays
      ) {
        parseInstallations.push(installation);
        parsed = true;
      }

      console.log(`${installation.account.login} has been checked ${
        parsedCache[installation.account.login].offsetDays} days ago, ${parsed ? 'adding to the queue' : 'skipping'}`)
    }
  }

  for await (const installation of parseInstallations) {
    if (parseData.length >= limitUsers) {
      break
    }

    console.log(`Fetching stars data for user ${installation.account.login}`)

    // check if we have access to open-sauced-goals repo and fail safe exit if not
    const installationExists = await app.octokit.rest.apps
      .getRepoInstallation({
        owner: installation.account.login,
        repo: 'open-sauced-goals',
      })
      .then(() => true)
      .catch((response) => {
        if (response.status === 404) {
          console.log(`${installation.account.login} is missing open-sauced-goals repo, flagging to skip`)
          checked[installation.account.login] = {
            owner: installation.account.login,
            notFound: true,
            lastExecuted,
          }
        } else {
          console.log(`Error getting repo installation for ${installation.account.login}`, response.toString())
        }

        return false
      })

    // if installation exists we proceed towards parsing
    if (installationExists) {
      const octokit = await app.getInstallationOctokit(installation.id);

      const {data: repository} = await octokit.rest.repos.get({
        owner: installation.account.login,
        repo: 'open-sauced-goals',
      }).catch((err) => {
        console.log(`${installation.account.login} open-sauced-goals: ${err}`)
        return {data: null}
      })

      const {data} = await octokit.rest.repos.getContent({
        owner: installation.account.login,
        repo: 'open-sauced-goals',
        path: 'stars.json'
      }).catch((err) => {
        console.log(`${installation.account.login} stars.json: ${err}`)
        return {data: {content: btoa("[]")}}
      })

      // convert from base64 to parseable JSON
      const content = Buffer.from(data.content, "base64").toString()
      const parsedData = JSON.parse(content);
      const starsData = data.content.length > 0

      parseData.push({
        installation,
        parsedData,
        starsData,
        repository,
      })
    }
  }

  consoleHeader('Parsing stars');
  console.log(`Existing installation queue was ${parseInstallations.length}`)
  console.log(`Attempting to parse ${parseData.length} users out of ${limitUsers} process.env.LIMIT_USERS`)

  await Promise.all(
    parseData.map(async ({installation, parsedData, starsData, repository}) =>
      new Promise(async (resolve, reject) => {
        try {
          console.log(`Processing installation #${installation.id}, ${installation.account.login} stars.json`)
          const octokit = await app.getInstallationOctokit(installation.id);

          // update data with repo id
          for await (const item of parsedData) {
            const [owner, repo] = item.full_name.split("/")
            const currentRepoResponse = await octokit.rest.repos.get({owner, repo})
              .catch((err) => {
                return {errors: [err]}
              })

            if (currentRepoResponse.errors && currentRepoResponse.errors.length > 0) {
              console.log(`ERROR for ${owner}/${repo}`, currentRepoResponse.errors)
              // reject(currentRepoResponse.errors)
              continue
            }

            const {
              id,
              stargazers_count,
              description,
              open_issues,
            } = currentRepoResponse.data

            const {data, errors} = await api.persistedRepoDataFetch({owner, repo})

            if (errors && errors.length > 0) {
              console.log(`ERROR for ${owner}/${repo}`, errors)
              // reject(errors)
              continue
            }

            if (
              data.gitHub.repositoryOwner === null
              || typeof data.gitHub.repositoryOwner.repository !== "object"
            ) {
              console.log(`ERROR for ${owner}/${repo}`, "No owner")
              // reject("No owner")
              continue
            }

            const {contributors_oneGraph} = data.gitHub.repositoryOwner.repository;

            const contributorNames = await fetchContributorNames(contributors_oneGraph.nodes)

            item.id = id

            const userStars = {
              id: repository.id,
              user_id: installation.account.id,
              star_id: item.id,
              repo_name: item.full_name,
              recency_score: parsedData.indexOf(item),
              description: description,
              issues: open_issues,
              stars: stargazers_count,
              contributors: contributorNames.slice(0,2) // grab first two names only
            };

            await supabase
              .from('user_stars')
              .insert(userStars)
          }

          // send parsedData to stars table
          await supabase
            .from('stars')
            .upsert(parsedData, {
              onConflict: "id"
            })

          console.log(`ADDED STARS FROM: ${installation.account.login}`)

          // send parsedData to supabase
          supabase
            .from('users')
            .upsert({
              id: installation.account.id,
              stars_data: starsData,
              open_issues: repository.open_issues_count,
              private: repository.private,
            }, {
              onConflict: "id"
            })

          checked[installation.account.login] = {
            owner: installation.account.login,
            notFound: false,
            lastExecuted,
          }

          resolve(checked[installation.account.login]);
        } catch (err) {
          console.log(`UNEXPECTED ERROR: ${err}`)
          console.log(`SKIPPED: ${installation.account.login}`)
          reject(err);
        }
      })
    )
  )

  // check whether we have new data to cache
  consoleHeader('Versioning changes');
  if (parseData.length > 0) {
    console.log('cron.json cached users: ', Object.keys(cron.checked).length);
    console.log('cron.json parsed users: ', parseData.length);

    // write to file and commit block
    await writeFile('./src/cron.json', JSON.stringify({
      lastExecuted,
      checked
    }, null, 2))
    console.log('Wrote changes to cron.json, make sure to commit this file');
  } else {
    console.log('Nothing to commit to cron.json');
  }

  consoleHeader('Finished');
}

await run()


src/contributors.js
import consoleHeader from './lib/consoleHeader.js'
import api from "./lib/persistedGraphQL.js";
import fetchContributorNames from "./lib/contributorNameHelper.js";
import {p} from "@antfu/utils";
import {supabase} from "./lib/supabase.js";

const limitDays = parseInt(process.env.LIMIT_CONTRIBUTOR_DAYS) || 5
const lastExecuted = new Date()

consoleHeader(' OPEN |SAUCED', {
  font: 'block',
})
console.log(`Started execution at ${lastExecuted}`)

async function run() {
  const testDate = new Date(lastExecuted.getTime() - (limitDays * 24 * 60 * 60 * 1000))

  const {data: getAvailableRepo, error} = await supabase
    .from('repos')
    .select('*')
    .lte('last_fetched_contributors_at', JSON.stringify(testDate))
    .limit(1)
    .single()

  if (error) {
    console.log(`Unable to fetch repos from supabase`, error)
    return
  }

  const [owner, repo] = getAvailableRepo.full_name.split('/');
  const {data, errors} = await api.persistedRepoDataFetch({owner, repo})

  if (errors && errors.length > 0) {
    console.log(`ERROR for ${owner}/${repo}`, errors)
    return
  }

  if (
    data.gitHub.repositoryOwner === null
    || typeof data.gitHub.repositoryOwner.repository !== "object"
  ) {
    console.log(`ERROR for ${owner}/${repo}`, "No owner")
    return
  }

  const {contributors_oneGraph} = data.gitHub.repositoryOwner.repository

  const contributorNames = await fetchContributorNames(contributors_oneGraph.nodes)

  const contributions = []

  await p(contributorNames.slice(0, 10))
    .map(async (contributor) => {
      const query = `repo:${owner}/${repo} type:pr is:merged author:${contributor}`;
      const {data, errors} = await api.persistedGitHubContributions({query});
      const prCount = typeof data.gitHub.search.nodes !== "undefined" && data.gitHub.search.nodes.length || 0;

      if (errors && errors.length > 0) {
        console.log(`Error executing persistedQuery: ${query}`, errors);
        return;
      }

      const contribution = {
        repo_id: getAvailableRepo.id,
        contributor,
        count: prCount,
        last_merged_at: data.gitHub.search.nodes[0].mergedAt,
        url: data.gitHub.search.nodes[0].url,
      }

      console.log(contribution)
      process.exit(1)

      if (prCount > 0) {
        contributions.push(contribution)

        return contribution
      }
    })

  console.log(contributions);

  // console.log(data);
  // console.log(errors);
  process.exit(1);
}

await run()


src/lib/supabase.js
import { open, readFile, writeFile } from 'node:fs/promises'
import path from 'node:path'
import { createClient } from '@supabase/supabase-js'
import { p } from '@antfu/utils'
import { stringify } from 'csv-stringify'

const anon_key = process.env.SUPABASE_ANON_KEY
const supabaseUrl = process.env.SUPABASE_URL

const supabase = createClient(supabaseUrl, anon_key)

const supaCount = async (table, field = '*') => supabase
    .from(table)
    .select(field, {
      head: false,
      count: 'exact'
    })
    .range(0, 1)

const supaDump = async (basePath, table, columns = [], rows = []) => {
  const timestamp = new Date().toISOString()
  const filePath = path.join(basePath, `${table}.sql`)
  const csvPath = path.join(basePath, `${table}.csv`)
  const finalRow = rows.pop()
  const file = await open(new URL(`../${filePath}`, import.meta.url).pathname, 'w')

  await file.write(`--
-- Data for Name: ${table}; Type: TABLE DATA; Schema: public; Timestamp: ${timestamp}
-- COPY ${table}(${columns.join(', ')}) FROM '${table}.csv' WITH DELIMITER ',' CSV HEADER;
--

INSERT INTO ${table}(${columns.join(', ')}) VALUES
`)

  await p(rows)
    .map(async row => file.write(`(${columns.map(col => JSON.stringify(row[col]))
      .join(', ')
      .replaceAll('\\"', "'")
      .replaceAll("\\'", "'")
      .replaceAll("'", "''")
      .replaceAll('"', "'")}),\n`))
    .then(() => file.write(`(${columns.map(col => JSON.stringify(finalRow[col]))
      .join(', ')
      .replaceAll('\\"', "'")
      .replaceAll("\\'", "'")
      .replaceAll("'", "''")
      .replaceAll('"', "'")});\n`))

  await file.close()

  await stringify(rows, {
    header: true,
    cast: {
      object: value => Array.isArray(value) ? `{"${value.join('", "')}"}` : JSON.stringify(value)
    }
  }, (err, data) =>
    writeFile(new URL(`../${csvPath}`, import.meta.url), data))

  return {
    filePath,
    timestamp,
  }
}

const supaSeed = async (seeds) => {
  const file = await open(new URL(`../../supabase/seed.sql`, import.meta.url), 'w');

  await p(seeds)
    .map(async ({filePath}) => {
      console.log(`Seeding ${filePath} into supabase local database`)
      const data = await readFile(new URL(`../${filePath}`, import.meta.url), 'utf8')
      await file.write(data)
    })
    .then(() => file.close())
}

export {
  supabase,
  supaCount,
  supaDump,
  supaSeed,
}


src/populate.js
import { App } from 'octokit'
import { writeFile } from 'node:fs/promises'
import { p } from '@antfu/utils'

import { supabase } from './lib/supabase.js'
import api from './lib/persistedGraphQL.js'
import fetchContributorNames from './lib/contributorNameHelper.js'
import consoleHeader from './lib/consoleHeader.js'
import cron from './cron.json' assert { type: 'json' }

const limitDays = parseInt(process.env.LIMIT_DAYS) || 1
let limitUsers = parseInt(process.env.LIMIT_USERS) || 5
const checked = {...cron.checked}
const lastExecuted = new Date()
const parsedCache = {}
const parseInstallations = []
const parseData = []

consoleHeader(' OPEN |SAUCED', {
  font: 'block',
})
console.log(`Started execution at ${lastExecuted}`)

for (const item in checked) {
  const date = new Date(checked[item].lastExecuted)
  const diff = lastExecuted - date

  // generate a cache of items with offset dates
  parsedCache[item] = {
    ...checked[item],
    offsetDays: Number(diff / 1000 / 60 / 60 / 24).toFixed(0)
  }
}

async function run() {
  const app = new App({
    appId: +process.env.OPEN_SAUCED_APP_ID,
    privateKey: process.env.OPEN_SAUCED_PRIVATE_KEY,
  })

  consoleHeader('Parsing installations')
  for await (const {installation} of app.eachInstallation.iterator()) {
    if (installation.account.login === 'open-sauced') {
      continue
    }

    console.log(`Fetching data for installation ${installation.account.login}`)
    // we don't have a local commit log of checking these installations
    if (typeof parsedCache[installation.account.login] === 'undefined') {
      console.log(`${installation.account.login} is not in our cache, adding to the queue`)
      parseInstallations.push(installation)
    } else {
      // we have a local cache of this installation
      let parsed = false

      if (
        parsedCache[installation.account.login].offsetDays >= limitDays
      ) {
        parseInstallations.push(installation)
        parsed = true
      }

      console.log(`${installation.account.login} has been checked ${
        parsedCache[installation.account.login].offsetDays} days ago, ${parsed ? 'adding to the queue' : 'skipping'}`)
    }
  }

  for await (const installation of parseInstallations) {
    if (parseData.length >= limitUsers) {
      break
    }

    console.log(`Fetching stars data for user ${installation.account.login}`)

    // check if we have access to open-sauced-goals repo and fail safe exit if not
    const installationExists = await app.octokit.rest.apps
      .getRepoInstallation({
        owner: installation.account.login,
        repo: 'open-sauced-goals',
      })
      .then(() => true)
      .catch((response) => {
        if (response.status === 404) {
          console.log(`${installation.account.login} is missing open-sauced-goals repo, flagging to skip`)
          checked[installation.account.login] = {
            owner: installation.account.login,
            notFound: true,
            lastExecuted,
          }
        } else {
          console.log(`Error getting repo installation for ${installation.account.login}`, response.toString())
        }

        return false
      })

    // if installation exists we proceed towards parsing
    if (installationExists) {
      const octokit = await app.getInstallationOctokit(installation.id)

      const {data: repository} = await octokit.rest.repos.get({
        owner: installation.account.login,
        repo: 'open-sauced-goals',
      }).catch((err) => {
        console.log(`${installation.account.login} open-sauced-goals: ${err}`)
        return {data: null}
      })

      const {data} = await octokit.rest.repos.getContent({
        owner: installation.account.login,
        repo: 'open-sauced-goals',
        path: 'stars.json'
      }).catch((err) => {
        console.log(`${installation.account.login} stars.json: ${err}`)
        return {data: {content: btoa("[]")}}
      })

      // convert from base64 to parseable JSON
      const content = Buffer.from(data.content, "base64").toString()
      const parsedData = JSON.parse(content)
      const starsData = data.content.length > 0

      parseData.push({
        installation,
        parsedData,
        starsData,
        repository,
      })
    }
  }

  consoleHeader('Parsing stars')
  console.log(`Existing installation queue was ${parseInstallations.length}`)
  console.log(`Attempting to parse ${parseData.length} users out of ${limitUsers} process.env.LIMIT_USERS`)

  await Promise.all(
    parseData.map(async ({installation, parsedData, starsData, repository}) =>
      new Promise(async (resolve, reject) => {
        try {
          console.log(`Processing installation #${installation.id}, ${installation.account.login} stars.json`)
          const octokit = await app.getInstallationOctokit(installation.id)

          // update data with repo id
          for await (const item of parsedData) {
            const [owner, repo] = item.full_name.split("/")
            const currentRepoResponse = await octokit.rest.repos.get({owner, repo})
              .catch((err) => {
                return {errors: [err]}
              })

            if (currentRepoResponse.errors && currentRepoResponse.errors.length > 0) {
              console.log(`ERROR for ${owner}/${repo}`, currentRepoResponse.errors)
              // reject(currentRepoResponse.errors)
              continue
            }

            const {
              id,
              stargazers_count,
              description,
              open_issues,
            } = currentRepoResponse.data

            const {data, errors} = await api.persistedRepoDataFetch({owner, repo})

            if (errors && errors.length > 0) {
              console.log(`ERROR for ${owner}/${repo}`, errors)
              // reject(errors)
              continue
            }

            if (
              data.gitHub.repositoryOwner === null
              || typeof data.gitHub.repositoryOwner.repository !== "object"
            ) {
              console.log(`ERROR for ${owner}/${repo}`, "No owner")
              // reject("No owner")
              continue
            }

            const {contributors_oneGraph} = data.gitHub.repositoryOwner.repository

            const contributorNames = await fetchContributorNames(contributors_oneGraph.nodes)

            const contributions = []

            await p(contributorNames.slice(0, 10))
              .map(async (contributor) => {
                const query = `repo:${owner}/${repo} type:pr is:merged author:${contributor}`;
                const {data, errors} = await api.persistedGitHubContributions({query});
                const prCount = typeof data.gitHub.search.nodes !== "undefined" && data.gitHub.search.nodes.length || 0;

                if (errors && errors.length > 0) {
                  console.log(`Error executing persistedQuery: ${query}`, errors);
                  return;
                }

                if (prCount > 0) {
                  return contributions.push({
                    name: contributor,
                    count: prCount,
                    lastMergedAt: data.gitHub.search.nodes[0].mergedAt,
                    url: data.gitHub.search.nodes[0].url,
                  });
                }
              })

            item.id = id

            const userStars = {
              id: repository.id,
              user_id: installation.account.id,
              star_id: item.id,
              repo_name: item.full_name,
              recency_score: parsedData.indexOf(item),
              description: description,
              issues: open_issues,
              stars: stargazers_count,
              contributors: contributorNames.slice(0,2), // grab first two names only
              // don't limit the json field for now
              // unsure of final format or sort order
              contributions: contributions.sort((a, b) => a.count > b.count ? -1 : 1),
            }

            await supabase
              .from('user_stars')
              .upsert(userStars, {
                onConflict: "id"
              })
          }

          // send parsedData to stars table
          await supabase
            .from('stars')
            .upsert(parsedData, {
              onConflict: "id"
            })

          console.log(`ADDED STARS FROM: ${installation.account.login}`)

          // send parsedData to supabase
          await supabase
            .from('users')
            .upsert({
              id: installation.account.id,
              login: installation.account.login,
              stars_data: starsData,
              open_issues: repository.open_issues_count,
              private: repository.private,
            }, {
              onConflict: "id"
            })

          checked[installation.account.login] = {
            owner: installation.account.login,
            notFound: false,
            lastExecuted,
          }

          resolve(checked[installation.account.login])
        } catch (err) {
          console.log(`UNEXPECTED ERROR: ${err}`)
          console.log(`SKIPPED: ${installation.account.login}`)
          reject(err)
        }
      })
    )
  )

  // check whether we have new data to cache
  consoleHeader('Versioning changes')
  if (parseData.length > 0) {
    console.log('cron.json cached users: ', Object.keys(cron.checked).length)
    console.log('cron.json parsed users: ', parseData.length)

    // write to file and commit block
    await writeFile('./src/cron.json', JSON.stringify({
      lastExecuted,
      checked
    }, null, 2))
    console.log('Wrote changes to cron.json, make sure to commit this file')
  } else {
    console.log('Nothing to commit to cron.json')
  }

  consoleHeader('Finished')
}

await run()
